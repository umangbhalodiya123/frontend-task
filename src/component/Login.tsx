'use client'
import { loginAction, selectStatus } from '@/store/auth/authSlice'
import { useAppDispatch, useAppSelector } from '@/store/hooks'
import { LockOutlined, Visibility, VisibilityOff } from '@mui/icons-material'
import {
  Avatar,
  Box,
  Button,
  CircularProgress,
  Container,
  FormControl,
  IconButton,
  InputAdornment,
  InputLabel,
  OutlinedInput,
  TextField,
  Typography,
} from '@mui/material'
import { useState } from 'react'

export const Login = () => {
  const dispatch = useAppDispatch()
  const status = useAppSelector(selectStatus)
  const [showPassword, setShowPassword] = useState(false)

  const onSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault()
    const formData = new FormData(event.currentTarget)
    const data = {
      email: formData.get('email') as string,
      password: formData.get('password') as string,
    }
    if (data.email && data.password) {
      dispatch(loginAction(data))
    }
  }

  return (
    <Container component="main" maxWidth="xs" aria-label="login">
      <Box
        sx={{
          marginTop: 8,
          display: 'flex',
          flexDirection: 'column',
          alignItems: 'center',
        }}
      >
        <Avatar sx={{ m: 1, bgcolor: 'secondary.main' }}>
          <LockOutlined />
        </Avatar>
        <Typography component="h1" variant="h5">
          Login
        </Typography>
        <Box component="form" onSubmit={onSubmit} noValidate sx={{ mt: 1 }}>
          <TextField
            margin="normal"
            required
            fullWidth
            id="email"
            label="Email Address"
            name="email"
            autoFocus
          />

          <FormControl margin="normal" variant="outlined" fullWidth>
            <InputLabel htmlFor="password">Password</InputLabel>
            <OutlinedInput
              id="password"
              name="password"
              type={showPassword ? 'text' : 'password'}
              endAdornment={
                <InputAdornment position="end">
                  <IconButton
                    aria-label="toggle password visibility"
                    onClick={() => setShowPassword((prev) => !prev)}
                    edge="end"
                  >
                    {showPassword ? <VisibilityOff /> : <Visibility />}
                  </IconButton>
                </InputAdornment>
              }
              label="Password"
            />
          </FormControl>

          <Button
            type="submit"
            fullWidth
            variant="contained"
            sx={{ mt: 3, mb: 2 }}
            disabled={status === 'loading'}
            startIcon={status === 'loading' && <CircularProgress size={16} />}
          >
            {status === 'loading' ? 'Logging in...' : 'Login'}
          </Button>
        </Box>
      </Box>
    </Container>
  )
}
