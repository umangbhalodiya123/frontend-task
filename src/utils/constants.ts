import { Dashboard, Info, Person, Settings } from '@mui/icons-material'

export const API_URL = 'https://api.mockaroo.com/api/502602f0'

export const API_KEY = process.env.NEXT_PUBLIC_API_KEY

export const CREDENTIALS = [
  {
    id: '1',
    name: 'John Doe',
    email: 'admin@mail.com',
    password: 'Admin@123',
  },
]

export const NAVBAR_WIDTH = 240

export const NAVBAR_ITEMS = [
  {
    name: 'Dashboard',
    path: '/',
    icon: Dashboard,
  },
  {
    name: 'About',
    path: '/about',
    icon: Info,
  },
  {
    name: 'Profile',
    path: '/profile',
    icon: Person,
  },
  {
    name: 'Settings',
    path: '/settings',
    icon: Settings,
  },
]
