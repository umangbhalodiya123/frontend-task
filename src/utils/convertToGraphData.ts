import { Employee } from '@/types'

export const convertToGraphData = (data: Employee[], key: keyof Employee) => {
  const graphData: { label: string; value: number }[] = []

  const labels = [...new Set(data.map((employee) => employee[key]))].toSorted(
    (a, b) =>
      typeof a === 'number' && typeof b === 'number'
        ? a - b
        : typeof a === 'string' && typeof b === 'string'
        ? a.localeCompare(b)
        : 0
  )

  labels.forEach((label) => {
    const avgPerformanceRating =
      data
        .filter((employee) => employee[key] === label)
        .reduce((acc, curr) => acc + curr.performance_rating, 0) /
      data.filter((employee) => employee[key] === label).length

    graphData.push({
      label: String(label),
      value: avgPerformanceRating,
    })
  })

  return graphData
}
