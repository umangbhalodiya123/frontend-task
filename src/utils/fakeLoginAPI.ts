import { LoginRequestValues, User } from '../types'
import { CREDENTIALS } from '.'

export const login = (values?: LoginRequestValues, timeout = 700) => {
  return new Promise<{ success: boolean; user?: User }>((resolve, reject) => {
    setTimeout(() => {
      const user = CREDENTIALS.find(
        (credential) =>
          credential.email === values?.email &&
          credential.password === values?.password
      )
      if (user) {
        resolve({ success: true, user })
      } else {
        reject({ success: false })
      }
    }, timeout)
  })
}
