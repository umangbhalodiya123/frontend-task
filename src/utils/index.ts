export * from './theme'
export * from './constants'
export * from './storage'
export * from './convertToGraphData'
export * from './fakeLoginAPI'
export * from './mocks'
