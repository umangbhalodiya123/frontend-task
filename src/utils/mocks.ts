export const employees = [
  {
    id: 1,
    name: 'Ruby Goadby',
    gender: 'Male',
    age: 24,
    job_title: 'Teacher',
    department: 'Sales',
    performance_rating: 5,
    attendance_percentage: 87.87,
    overtime_hours: 7.55,
    team_size: 11,
    communication_skills: 'good',
    employee_status: 'on leave',
  },
  {
    id: 2,
    name: 'Terencio Feeney',
    gender: 'Male',
    age: 24,
    job_title: 'Social Worker',
    department: 'Legal',
    performance_rating: 4,
    attendance_percentage: 87.2,
    overtime_hours: 19.8,
    team_size: 3,
    communication_skills: 'average',
    employee_status: 'active',
  },
  {
    id: 3,
    name: 'Orazio Libbis',
    gender: 'Male',
    age: 23,
    job_title: 'Nurse Practicioner',
    department: 'Services',
    performance_rating: 3,
    attendance_percentage: 96.54,
    overtime_hours: 15.61,
    team_size: 9,
    communication_skills: 'good',
    employee_status: 'active',
  },
  {
    id: 4,
    name: 'Meara Jewell',
    gender: 'Female',
    age: 26,
    job_title: 'Analyst Programmer',
    department: 'Research and Development',
    performance_rating: 8,
    attendance_percentage: 36.57,
    overtime_hours: 5.66,
    team_size: 7,
    communication_skills: 'good',
    employee_status: 'active',
  },
  {
    id: 5,
    name: 'Farlee Solomonides',
    gender: 'Male',
    age: 23,
    job_title: 'Human Resources Assistant IV',
    department: 'Business Development',
    performance_rating: 7,
    attendance_percentage: 57.37,
    overtime_hours: 17.96,
    team_size: 9,
    communication_skills: 'poor',
    employee_status: 'active',
  },
  {
    id: 6,
    name: 'Brandice Kilbee',
    gender: 'Female',
    age: 21,
    job_title: 'VP Sales',
    department: 'Product Management',
    performance_rating: 2,
    attendance_percentage: 31.24,
    overtime_hours: 15.78,
    team_size: 6,
    communication_skills: 'good',
    employee_status: 'active',
  },
  {
    id: 7,
    name: 'Helsa Grealy',
    gender: 'Agender',
    age: 39,
    job_title: 'Recruiting Manager',
    department: 'Training',
    performance_rating: 9,
    attendance_percentage: 73.98,
    overtime_hours: 13.99,
    team_size: 15,
    communication_skills: 'excellent',
    employee_status: 'active',
  },
  {
    id: 8,
    name: 'Kimmie Bowart',
    gender: 'Female',
    age: 23,
    job_title: 'Assistant Professor',
    department: 'Services',
    performance_rating: 10,
    attendance_percentage: 41.31,
    overtime_hours: 19.8,
    team_size: 3,
    communication_skills: 'average',
    employee_status: 'active',
  },
  {
    id: 9,
    name: 'Desmond Bligh',
    gender: 'Male',
    age: 31,
    job_title: 'Software Test Engineer II',
    department: 'Services',
    performance_rating: 4,
    attendance_percentage: 4.27,
    overtime_hours: 4.57,
    team_size: 9,
    communication_skills: 'poor',
    employee_status: 'on leave',
  },
  {
    id: 10,
    name: 'Dana Schuelcke',
    gender: 'Male',
    age: 37,
    job_title: 'Safety Technician II',
    department: 'Training',
    performance_rating: 5,
    attendance_percentage: 42.8,
    overtime_hours: 19.09,
    team_size: 17,
    communication_skills: 'average',
    employee_status: 'on leave',
  },
]
