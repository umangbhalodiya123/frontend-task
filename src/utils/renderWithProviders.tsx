import React, { PropsWithChildren } from 'react'
import { render } from '@testing-library/react'
import { Provider } from 'react-redux'
import { store } from '@/store'

export function renderWithProviders(ui: React.ReactElement) {
  const Wrapper = ({ children }: PropsWithChildren) => (
    <Provider store={store}>{children}</Provider>
  )

  return {
    store,
    ...render(ui, { wrapper: Wrapper }),
  }
}
