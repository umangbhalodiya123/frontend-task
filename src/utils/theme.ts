'use client'
import { Roboto } from 'next/font/google'
import { createTheme } from '@mui/material/styles'
import { deepOrange, orange } from '@mui/material/colors'

const roboto = Roboto({
  weight: ['300', '400', '500', '700'],
  subsets: ['latin'],
  display: 'swap',
})

export const theme = createTheme({
  palette: {
    mode: 'dark',
    primary: deepOrange,
    secondary: orange,
  },
  typography: {
    fontFamily: roboto.style.fontFamily,
  },
})
