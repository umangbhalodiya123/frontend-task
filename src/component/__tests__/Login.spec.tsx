import { screen } from '@testing-library/react'
import { Login } from '../Login'
import { renderWithProviders } from '@/utils/renderWithProviders'

describe('Login', () => {
  it('renders correctly', () => {
    renderWithProviders(<Login />)

    const loginContainer = screen.getByLabelText('login')
    expect(loginContainer).toBeInTheDocument()
  })

  it('renders matches snapshot', () => {
    renderWithProviders(<Login />)

    const loginContainer = screen.getByLabelText('login')
    expect(loginContainer).toBeInTheDocument()
    expect(loginContainer).toMatchSnapshot()
  })
})
