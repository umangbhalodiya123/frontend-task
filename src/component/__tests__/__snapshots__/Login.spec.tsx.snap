// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders matches snapshot 1`] = `
<main
  aria-label="login"
  class="MuiContainer-root MuiContainer-maxWidthXs css-nzinn5-MuiContainer-root"
>
  <div
    class="MuiBox-root css-binzgt"
  >
    <div
      class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-lguxed-MuiAvatar-root"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="LockOutlinedIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2M9 6c0-1.66 1.34-3 3-3s3 1.34 3 3v2H9zm9 14H6V10h12zm-6-3c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2"
        />
      </svg>
    </div>
    <h1
      class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
    >
      Login
    </h1>
    <form
      class="MuiBox-root css-164r41r"
      novalidate=""
    >
      <div
        aria-label="email"
        class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-focused Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for="email"
          id="email-label"
        >
          Email Address
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id="email"
            name="email"
            required=""
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-14lo706"
            >
              <span>
                Email Address
                 
                *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth css-1iledgx-MuiFormControl-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="password"
        >
          Password
        </label>
        <div
          aria-label="password"
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
            id="password"
            name="password"
            type="password"
            value=""
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
          >
            <button
              aria-label="toggle password visibility"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="VisibilityIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5M12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5m0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Password
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth css-1vhaqj4-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        Login
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </form>
  </div>
  <div
    class="MuiBox-root css-0"
  >
    <div
      class="MuiStack-root css-nl1xw3-MuiStack-root"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-1hp3b1n-MuiTypography-root"
      >
        Hint
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      >
        Email: admin@mail.com
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      >
        Password: Admin@123
      </p>
    </div>
  </div>
</main>
`;
